devices/flash-flexspi/fspi/fspi.c

```diff
@@ -0,0 +1,476 @@
+#include "fspi_rt105x.h"
+#else
+#error "FlexSPI is not supported on this target"
+#endif
+
+
+/* Stop/Start FlexSPI module */
+__attribute__((section(".noxip"))) static void flexspi_disable(flexspi_t *fspi, int disable)
+{
+	register u32 reg = *(fspi->base + mcr0);
+
+	if (disable) {
+		reg |= (1 << 1);
+	}
+	else {
+		reg &= ~(1 << 1);
+	}
+
+	*(fspi->base + mcr0) = reg;
+}
+
+
+/* Module Software Reset (all registers except control registers) */
+__attribute__((section(".noxip"))) static inline void flexspi_swreset(flexspi_t *fspi)
+{
+	*(fspi->base + mcr0) |= 1;
+	while ((*(fspi->base + mcr0) & 1) != 0)
+		;
+}
+
+
+__attribute__((section(".noxip"))) void flexspi_lutUpdate(flexspi_t *fspi, u32 index, const u32 *lutTable, size_t count)
+{
+	u32 lutCopy[64];
+	unsigned int i;
+	volatile u32 *lutPtr = fspi->base + lut64 + index;
+
+	for (i = 0; i < count; ++i) {
+		lutCopy[i] = *lutTable++;
+	}
+
+	/* Wait for bus and lut sequencer to be idle */
+	while (((*(fspi->base + sts0) & 2) != 0 && (*(fspi->base + sts0) & 1) != 0) == 0)
+		;
+
+	/* Unlock LUT */
+	*(fspi->base + lutkey) = 0x5af05af0;
+	*(fspi->base + lutcr) = 2;
+
+	/* Update LUT */
+	for (i = 0; i < count; ++i) {
+		*lutPtr++ = lutCopy[i];
+	}
+
+	/* Lock LUT */
+	*(fspi->base + lutkey) = 0x5af05af0;
+	*(fspi->base + lutcr) = 1;
+}
+
+
+__attribute__((section(".noxip"))) int flexspi_init(flexspi_t *fspi, int instance, u8 slPortMask)
+{
+	int res;
+	unsigned int i;
+	u32 reg, lut[4], *base = flexspi_getBase(instance);
+
+	if (base == NULL || (slPortMask & ~0xf) != 0 || (slPortMask & 0xf) == 0) {
+		return -EINVAL;
+	}
+
+	fspi->base = base;
+	fspi->instance = instance;
+	fspi->ahbAddr = flexspi_ahbAddr(instance);
+	fspi->slPortMask = slPortMask;
+
+	res = flexspi_pinConfig(fspi);
+	if (res < 0) {
+		return res;
+	}
+
+	hal_interruptsDisable();
+
+	hal_disableDCache();
+	hal_disableICache();
+
+	hal_invalDCacheAll();
+	hal_cleanDCache();
+
+	flexspi_clockConfig(fspi);
+
+	/* Release FlexSPI from reset and power SRAM */
+	flexspi_disable(fspi, 0);
+	flexspi_swreset(fspi);
+
+	/* Make sure controller is configured in module stop mode */
+	flexspi_disable(fspi, 1);
+
+	/* Set AHB and IP timeouts to default value do not enable MCR0[MDIS] module yet */
+	reg = *(fspi->base + mcr0) & ((1 << 15) | (3 << 2));
+	*(fspi->base + mcr0) = reg | (0xff << 24) | (0xff << 16) | (1 << 4);
+
+	/* Disable "same device" mode and configure all slave devices independently */
+	reg = *(fspi->base + mcr2);
+	*(fspi->base + mcr2) = reg & ~(1 << 15);
+
+	/* Enable AHB Prefetch, Bufferable, Read Prefetch & Read Address option */
+	*(fspi->base + ahbcr) |= (1 << 3) | (1 << 4) | (1 << 5) | (1 << 6);
+
+	/* Disable all AHB buffers */
+	for (i = 0; i < AHBRXBUF_CNT; ++i) {
+		*(fspi->base + ahbrxbuf0cr0 + i) = 0;
+	}
+
+#if 1
+	/* Set default all AHB buffers (many small buffers) with prefetch enabled */
+	for (i = 0; i < AHBRXBUF_CNT; ++i) {
+		*(fspi->base + ahbrxbuf0cr0 + i) = (1 << 31) | ((i & 7) << 16) | (1 << 6);
```

Comment:
I'm surprised this didn't throw a warning (integer overflow)
```suggestion
		*(fspi->base + ahbrxbuf0cr0 + i) = (1U << 31) | ((i & 7) << 16) | (1 << 6);
```

Reply:
That's right corrected also `(0xff << 24)` and other shifts that cannot be represented in type `int`.